### ISSUE TITLE
SQL syntax highlighting inside sqlx macros

### ISSUE THREAD

**Issue by Milo123459** [rust-analyzer/rust-analyzer#10448]
Something like SQL syntax-highlighting inside usage of SQLx macros. This could also provide autocomplete for columns in tables, along with SQL keywords like UPDATE, INSERT, etc.

**Comment by lnicola** [comment]
How do you see that working?

**Comment by Milo123459** [comment]
> How do you see that working?

There is autocomplete for SQL files, and countless extensions to do that if not provided on install of VSCode; literally that, but inside the SQLx query macro. And for the autocomplete, SQLx actually provides this. The offline feature can do this, but, if it needs to be on the fly a modification would be made to SQLx, or we write something that fetches the data in RA

**Comment by lnicola** [comment]
Unfortunately, we can't do that, for various technical reasons. And I don't think we should special-case `sqlx`, nor implement parsing for the three popular SQL dialects it supports.

**Comment by flodiebold** [comment]
Yeah. We already have #10281 and #7402 for the general solution.

**Comment by Milo123459** [comment]
Ah, I've done this wrong. I meant overall SQL syntax highlighting, but they look good.

**Comment by flodiebold** [comment]
So since we're not going to provide any specific SQL support, let's close this in favor of those issues.

**Comment by lnicola** [comment]
@Milo123459 it's a duplicate of #10281, filed also by you.

**Comment by Milo123459** [comment]
I am aware. I wanted this one to be more specific

**Comment by peri4n** [comment]
If you ran into this and use NeoVim. You can use Treesitter to highlight the SQL inside string literals by putting

```
(macro_invocation
  macro: (scoped_identifier
    path: (identifier) @_path (#eq? @_path "sqlx")
    name: (identifier) @_name (#any-of? @_name "query" "query_as")
  )
  (token_tree 
    (_ (string_content) @injection.content)
    (#set! injection.language "sql")
  )
)

(call_expression
  (scoped_identifier
    path: (identifier) @path (#eq? @path "sqlx")
    name: (identifier) @name (#any-of? @name "query" "query_as"))
  (arguments
    (_ (string_content) @injection.content)
    (#set! injection.language "sql")))
```

into `$USER/.config/neovim/queries/rust/injections.scm`

**Comment by Milo123459** [comment]
Is there an equivalent for vscode?

**Comment by peri4n** [comment]
I am not a VS Code expert but I don't believe there is. Treesitter support is an open issue: https://github.com/microsoft/vscode/issues/50140

**Comment by paulwongx** [comment]
To clarify, based on this `issue` is it normal to see SQL queries inside a `sqlx::query_as!` macro all in one color like below?

![Image](https://github.com/user-attachments/assets/86cf9bfe-f1b0-46ff-9e95-9c9dd791138b)

**Comment by LeonSnajdr** [comment]
> Is there an equivalent for vscode?

You can use https://marketplace.visualstudio.com/items?itemName=qufiwefefwoyn.inline-sql-syntax 

Then it looks like this:
![Image](https://github.com/user-attachments/assets/725a7a2e-d7c9-4827-91fa-645fa53981ab)

**Comment by cies** [comment]
Another options is to use `include_str!()` to put the SQL in a file of it's own (not a bad idea for several reasons).

There's a problem with rust-analyzer and this macro: it only does CTRL-click to goto definition for `.rs` files. I found a work around I described here: https://github.com/rust-lang/rust-analyzer/issues/11025

Then you want to use a VSCode plugin to connect to Postgres in order to do proper syntax highlighting/ intellisense. There are several that can help with that:

https://marketplace.visualstudio.com/items?itemName=Supabase.postgrestools

https://marketplace.visualstudio.com/items?itemName=ckolkman.vscode-postgres

https://marketplace.visualstudio.com/items?itemName=mtxr.sqltools
